File: ebitengine/ui/button.go
package ui

import (
	"image/color"

	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/text/v2"
	"github.com/hajimehoshi/ebiten/v2/vector"
	"github.com/thesimpledev/radianthorizon/core"
)

// Button is a clickable UI button with hover and press states.
type Button struct {
	Text    string
	X, Y    float64
	W, H    float64
	OnClick func()
	Font    *text.GoTextFace

	Hovered bool
	Pressed bool
	Visible bool
}

// NewButton creates a button.
func NewButton(label string, x, y, w, h float64, font *text.GoTextFace, onClick func()) *Button {
	return &Button{
		Text:    label,
		X:       x,
		Y:       y,
		W:       w,
		H:       h,
		OnClick: onClick,
		Font:    font,
		Visible: true,
	}
}

// Draw renders the button.
func (b *Button) Draw(screen *ebiten.Image) {
	if !b.Visible {
		return
	}

	// Background
	var bgColor color.NRGBA
	if b.Hovered {
		bgColor = core.ColorButtonHov
	} else {
		bgColor = core.ColorButton
	}
	vector.DrawFilledRect(screen, float32(b.X), float32(b.Y), float32(b.W), float32(b.H), bgColor, true)

	// Border
	vector.StrokeRect(screen, float32(b.X), float32(b.Y), float32(b.W), float32(b.H), 1, core.ColorWhite, true)

	// Text centered
	op := &text.DrawOptions{}
	tw, th := text.Measure(b.Text, b.Font, 0)
	op.GeoM.Translate(b.X+(b.W-tw)/2, b.Y+(b.H-th)/2)
	op.ColorScale.ScaleWithColor(core.ColorButtonTxt)
	text.Draw(screen, b.Text, b.Font, op)
}

// ContainsPoint checks if a point is inside the button.
func (b *Button) ContainsPoint(mx, my int) bool {
	fx, fy := float64(mx), float64(my)
	return fx >= b.X && fx <= b.X+b.W &&
		fy >= b.Y && fy <= b.Y+b.H
}

// HandleMouseMove updates hover state.
func (b *Button) HandleMouseMove(mx, my int) {
	if !b.Visible {
		return
	}
	b.Hovered = b.ContainsPoint(mx, my)
}

// HandleMousePress records a press.
func (b *Button) HandleMousePress(mx, my int) {
	if !b.Visible {
		return
	}
	if b.ContainsPoint(mx, my) {
		b.Pressed = true
	}
}

// HandleMouseRelease triggers the click if appropriate.
func (b *Button) HandleMouseRelease(mx, my int) {
	if !b.Visible {
		return
	}
	if b.Pressed && b.ContainsPoint(mx, my) {
		if b.OnClick != nil {
			b.OnClick()
		}
	}
	b.Pressed = false
}

File: love/ui/button.lua
-- ui/button.lua
-- A clickable button with hover and press visual states.

local Class = require("lib.class")
local C     = require("core.constants")

local Button = Class:extend()

function Button:init(text, x, y, w, h, on_click)
    self.text = text
    self.x = x
    self.y = y
    self.w = w
    self.h = h
    self.on_click = on_click -- callback function

    self.hovered = false
    self.pressed = false
    self.visible = true
end

function Button:draw()
    if not self.visible then return end

    -- Background color changes on hover
    if self.hovered then
        love.graphics.setColor(C.COLOR_BUTTON_HOV)
    else
        love.graphics.setColor(C.COLOR_BUTTON)
    end

    love.graphics.rectangle("fill", self.x, self.y, self.w, self.h, 6, 6)

    -- Border
    love.graphics.setColor(C.COLOR_WHITE)
    love.graphics.rectangle("line", self.x, self.y, self.w, self.h, 6, 6)

    -- Text (centered in the button)
    love.graphics.setColor(C.COLOR_BUTTON_TXT)
    local font = love.graphics.getFont()
    local text_w = font:getWidth(self.text)
    local text_h = font:getHeight()
    love.graphics.print(
        self.text,
        self.x + (self.w - text_w) / 2,
        self.y + (self.h - text_h) / 2
    )
end

function Button:contains_point(mx, my)
    return mx >= self.x and mx <= self.x + self.w
       and my >= self.y and my <= self.y + self.h
end

function Button:mousemoved(mx, my)
    if not self.visible then return end
    self.hovered = self:contains_point(mx, my)
end

function Button:mousepressed(mx, my, button)
    if not self.visible then return end
    if button == 1 and self:contains_point(mx, my) then
        self.pressed = true
    end
end

function Button:mousereleased(mx, my, button)
    if not self.visible then return end
    if button == 1 and self.pressed and self:contains_point(mx, my) then
        if self.on_click then
            self.on_click()
        end
    end
    self.pressed = false
end

return Button
